crates/bitwarden-ipc/src/ipc_client.rs

```diff
@@ -42,170 +47,270 @@ pub struct IpcClientTypedSubscription<Crypto, Com, Ses, Payload>
+    receiver: tokio::sync::broadcast::Receiver<IncomingMessage>,
     client: Arc<IpcClient<Crypto, Com, Ses>>,
     _payload: std::marker::PhantomData<Payload>,
 }
 
+#[derive(Debug, Error, Clone, PartialEq, Eq)]
+#[bitwarden_error(flat)]
+pub enum SubscribeError {
+    #[error("The IPC processing thread is not running")]
+    NotStarted,
+}
+
+#[derive(Debug, Error, PartialEq, Eq)]
+#[bitwarden_error(basic)]
+#[error("Failed to start the IPC client: {0}")]
+pub struct StartError(String);
+
+#[derive(Debug, Error, PartialEq, Eq)]
+#[bitwarden_error(flat)]
+pub enum ReceiveError {
+    #[error("Failed to subscribe to the IPC channel: {0}")]
+    Channel(#[from] tokio::sync::broadcast::error::RecvError),
+
+    #[error("Timed out while waiting for a message: {0}")]
+    Timeout(#[from] tokio::time::error::Elapsed),
+}
+
+#[derive(Debug, Error, PartialEq, Eq)]
+#[bitwarden_error(flat)]
+pub enum TypedReceiveError {
+    #[error("Failed to subscribe to the IPC channel: {0}")]
+    Channel(#[from] tokio::sync::broadcast::error::RecvError),
+
+    #[error("Timed out while waiting for a message: {0}")]
+    Timeout(#[from] tokio::time::error::Elapsed),
+
+    #[error("Typing error: {0}")]
+    Typing(String),
+}
+
+impl From<ReceiveError> for TypedReceiveError {
+    fn from(value: ReceiveError) -> Self {
+        match value {
+            ReceiveError::Channel(e) => TypedReceiveError::Channel(e),
+            ReceiveError::Timeout(e) => TypedReceiveError::Timeout(e),
+        }
+    }
+}
+
 impl<Crypto, Com, Ses> IpcClient<Crypto, Com, Ses>
 where
     Crypto: CryptoProvider<Com, Ses>,
     Com: CommunicationBackend,
-    Ses: SessionRepository<Session = Crypto::Session>,
+    Ses: SessionRepository<Crypto::Session>,
 {
     pub fn new(crypto: Crypto, communication: Com, sessions: Ses) -> Arc<Self> {
         Arc::new(Self {
             crypto,
             communication,
             sessions,
+
+            incoming: RwLock::new(None),
+            cancellation_handle: RwLock::new(None),
         })
     }
 
+    pub async fn start(self: &Arc<Self>) -> Result<(), StartError> {
+        let client = self.clone();
+        let (cancellation_handle_tx, mut cancellation_handle_rx) =
+            tokio::sync::watch::channel(false);
+        let (await_init_tx, await_init_rx) = tokio::sync::oneshot::channel();
+        let mut cancellation_handle = self.cancellation_handle.write().await;
+
+        *cancellation_handle = Some(cancellation_handle_tx);
+
+        let future = async move {
+            let com_receiver = client.communication.subscribe().await;
+            let (client_tx, client_rx) = tokio::sync::broadcast::channel(CHANNEL_BUFFER_CAPACITY);
+
+            let mut client_incoming = client.incoming.write().await;
+            *client_incoming = Some(client_rx);
+            drop(client_incoming);
+
+            await_init_tx
+                .send(())
+                .expect("Sending init signal should not fail");
+
+            loop {
```

Comment:
Is await_init_tx used to wait until the subscription is done? if so we could just subscribe outside the task.
```suggestion
    let com_receiver = client.communication.subscribe().await;
    let (client_tx, client_rx) = tokio::sync::broadcast::channel(CHANNEL_BUFFER_CAPACITY);

    let mut client_incoming = client.incoming.write().await;
    *client_incoming = Some(client_rx);
    drop(client_incoming);
        let future = async move {
            loop {
```

Reply:
Hmm, I'm not sure I follow, which task do you mean?

Reply:
I mean that currently in the code we create a future that we spawn in the background, but then we use a oneshot channel to synchronize and wait for initialization. If we instead do the initialization outside the task, we can avoid having to do that synchronization. With some pseudocode, this is what we currently have:

```rust
let (await_init_tx, await_init_rx) = tokio::sync::oneshot::channel();

// We create a future that gets spawned in the background
let future = async move {
    // Do some initialization
    some_init_stuff().await;

    // Notify that we're initialized
    await_init_tx.send(());

    // Process messages
    loop { ... }
};

// Spawn the task and...
spawn_future(future);

///  wait for the initialization to complete before continuing
await_init_rx.await;
```

if we pull the initialization code out of the future, then we can avoid the explicit syncrhonization:

```rust
// Do some initialization before the future is created, now we don't need to wait as it's already done
some_init_stuff().await;

// We create a future that gets spawned in the background
let future = async move {
    // Process messages
    loop { ... }
};

// Spawn the task and...
spawn_future(future);
```

Reply:
Fixed, much simpler!